examples/f16/components/fgintruder.py

```diff
@@ -0,0 +1,376 @@
+    format_string += 'I I'              # sim/multiplay/generic/bool[31] | BOOL_ARRAY (hex encoded array of 31 bools)
+    """
+    # header values
+    MAGIC = 0x46474653                  # "FGFS"
+    FG_MPP_VERSION = 0x00010001         # 1.1
+    MSG_ID = 0x00000007                 # positional data
+    REQ_RANGE_NUM = 0x000000a0          # visibility range in nm
+    REPLY_PORT = 0x00000000             # deprecated
+    DEFAULT_DELTA_T = 0.5
+    DEFAULT_FG_MP_PORT = 5001
+
+    def __init__(self, callsign: str, model_path: str, fallback_model_index: int):
+        """
+        * callsign -  first 7 chars of vehicle name
+        * model_path - path to the model, in the form of `Aircraft/f16/Models/F-16.xml`
+                       The base directory is assumed to be set with `--fg-aircraft`, typically
+                       set to `--fg-aircraft=/home/mpodhradsky/.fgfs/Aircraft/org.flightgear.fgaddon.stable_2020`
+        * fallback_model_index - AI model that should be used as a low-res substitute. Typical values:
+                      1 - c172
+                     56 - Cessna 337
+                    516 - F-16
+                    702 - air balloon
+        """
+        self.callsign = callsign[0:7].encode('utf-8')       
+        self.model_path = model_path.encode('utf-8')
+        self.fallback_model_index = fallback_model_index
+
+        # init vars
+        self.x = 0.0
+        self.y = 0.0
+        self.z = 0.0
+        self.x_ori = 0.0
+        self.y_ori = 0.0
+        self.z_ori = 0.0
+
+        self.phi = 0.0
+        self.theta = 0.0
+        self.psi = 0.0
+        self.pn = 0.0
+        self.pe = 0.0
+        self.pu = self.h0
+    
+    def update_and_send(self, inputs: typing.Optional[typing.List[float]] =None):
+        # update:
+        if inputs is not None:
+            self.phi = inputs[3]
+            self.theta = inputs[4]
+            self.psi = inputs[5]
+            self.pn = inputs[9]
+            self.pe = inputs[10]
+            self.pu = inputs[11]/self.FG_FT_IN_M
+
+        self.x, self.y, self.z, self.x_ori, self.y_ori, self.z_ori = Dubins2DConverter.convert_to_ecef(
+                                                                        self.pn, self.pe, self.pu,
+                                                                        self.phi, self.theta, self.psi,
+                                                                        self.lat0, self.lon0, self.h0
+                                                                        )
+
+        # time
+        cur_date = datetime.fromtimestamp(time.time(), timezone.utc)
+        self.time = (
+                cur_date.hour * 3600.0 + cur_date.minute * 60.0 + cur_date.second + cur_date.microsecond * 0.000001)
+
+        # TODO: what to do with control surfaces?
+
+        # send
+        self.sock.sendto(self.pack_to_struct(), (FlightGearBase.DEFAULT_FG_IP, self.DEFAULT_FG_MP_PORT))
+    
+    def get_format_string(self) -> str:
```

Comment:
No need to change this but something interesting: string concatenation has O(n^2) complexity because strings are immutable. An efficient alternative is to put all strings in a sequence and `''.join(my_array_of_strs)`

Reply:
Fixed